File: src/dotnet/AspireWorker/Sessions/SessionEvents.cs
using System.Text.Json.Serialization;
using JetBrains.Annotations;

namespace JetBrains.Rider.Aspire.Worker.Sessions;

[JsonDerivedType(typeof(ProcessStartedEvent))]
[JsonDerivedType(typeof(LogReceivedEvent))]
[JsonDerivedType(typeof(ProcessTerminatedEvent))]
internal interface ISessionEvent;

[PublicAPI]
internal sealed record ProcessStartedEvent(string SessionId, string NotificationType, long Pid)
    : ISessionEvent;

[PublicAPI]
internal sealed record LogReceivedEvent(string SessionId, string NotificationType, bool IsStdErr, string LogMessage)
    : ISessionEvent;

[PublicAPI]
internal sealed record ProcessTerminatedEvent(string SessionId, string NotificationType, int ExitCode)
    : ISessionEvent;

File: core/src/main/kotlin/com/jetbrains/rider/aspire/sessions/SessionEvents.kt
package com.jetbrains.rider.aspire.sessions

interface SessionEvent
data class SessionStarted(val id: String, val pid: Long) : SessionEvent
data class SessionTerminated(val id: String, val exitCode: Int) : SessionEvent
data class SessionLogReceived(val id: String, val isStdErr: Boolean, val message: String) : SessionEvent